openassessment/management/commands/upload_oa_data.py

```diff
@@ -96,12 +96,11 @@ def _dump_to_csv(self, course_id, csv_dir):
         Returns:
             None
         """
-        output_streams = {
-            name: open(os.path.join(csv_dir, rel_path), 'w')
```

Comment:
Pylint wanted `open` to happen inside a `with`.

Reply:
Is doing this multiple writers thing the same behavior as before? I assume we've got tests to verify that this still works as intended?

Reply:
I figured it would be the same but I am looking for someone better versed in Python to render an opinion. We are still doing multiple file `open`s (they used to be a for loop inside the output stream dict) but now each happens inside a context manager.

Reply:
That's not what I meant, that seems fine, I just mean wrt/ behavior of the CsvWriter?

Reply:
I think this might be the use-case for https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack

Reply:
While the multiple `open` calls themselves should be fine (each context manager should be finished upon each iteration), there's a possible performance regression here, as we're now calling `csv_writer.write_to_csv` 3x instead of just once (there are 3 `OUTPUT_CSV_PATHS`).

That method contains the following warning/comment [1]:

> NOTE: The current implementation optimizes for memory usage,
> but not for the number of database queries.  All the queries
> use indexed fields (the submission uuid), so they should be
> relatively quick.

I'm not familiar with the underlying code, but I'd suggest at least a shallow dive on the implications.

> I assume we've got tests to verify ...

Trust, but verify. Given that this modifies report generation (valuable to faculty/grading), I'd suggest manual confirmation of the output of the reports, before and after this change. Or at least, confirm there's actually sufficient test coverage. Especially given this repo's age/history, I wouldn't take this for granted.

- [1] https://github.com/edx/edx-ora2/blob/e755216bd5e524129cc6136dc22b38cfe71633ab/openassessment/data.py#L160-L163

Reply:
@jansenk , I'm pretty sure that's what it was doing: There was a `for` inside the `output_streams` which looked like it was expanding a dict of `file_name: file_resource` and preemptively opening all those files with write access.